Split mutate processor into one processor per function

javanna opened the pull request:
Split mutate processor into one processor per function

talevy commented on plugins/ingest/src/main/java/org/elasticsearch/ingest/processor/gsub/GsubProcessor.java:
```diff
@@ -0,0 +1,89 @@
+            List<Map<String, String>> gsubConfig = ConfigurationUtils.readList(config, "expressions");
+            List<GsubExpression> gsubExpressions = new ArrayList<>();
+            for (Map<String, String> stringObjectMap : gsubConfig) {
+                String field = stringObjectMap.get("field");
```
should we maybe make ConfigurationUtils more flexible to enable its use here? I understand that the error message may want to be customized in this case.

javanna replied:
not sure, I think it is ok to use ConfigurationUtils for the top level element and then have specific code in each factory, otherwise we need to add too much generic code to ConfigurationUtils to handle all the cases. We can always reconsider in the future though.

talevy commented on plugins/ingest/src/main/java/org/elasticsearch/ingest/processor/AbstractStringProcessor.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.List;
+import java.util.Map;
+
+public abstract class AbstractStringProcessor implements Processor {
```
I think the fact that a list with the key name of `fields` is hard-coded here, maybe we should give a more 
specific name to this? should we call this `AbstractStringsFieldProcessor`?

or we can force the factory to implement a `getStringsFieldName` or something of the sort to make the `fields` name to be required to be defined by each implementing processor factory.

javanna replied:
I think we can always change it when we need to. In general, trying to reuse this class further will make it more complicated (same reason why I didn't do it with other string processors like split and join, because their configuration is provided in a different format and it's not worthwhile to make this class even more generic). That is why I think it's ok as-is, it is not meant to be reused besides trim, uppercase and lowercase for now, we will see what comes up next though :)

talevy commented on plugins/ingest/src/test/java/org/elasticsearch/ingest/processor/AbstractStringProcessorTestCase.java:
```diff
@@ -0,0 +1,82 @@
+
+import static org.hamcrest.Matchers.equalTo;
+
+public abstract class AbstractStringProcessorTestCase extends ESTestCase {
```
should we add a test for unicode characters, so that we have strict expectations on how those should behave?

javanna replied:
yea we should

talevy commented:
just left a few questions, tests look great!

martijnvg commented on plugins/ingest/src/main/java/org/elasticsearch/ingest/processor/join/JoinProcessor.java:
```diff
@@ -0,0 +1,73 @@
+            if (list == null) {
+                throw new IllegalArgumentException("field [" + entry.getKey() + "] is null, cannot join.");
+            }
+            String joined = list.stream()
```
Just a note when back porting this piece of code needs to be changed. I think we should use java8 where possible and in this case back porting is trivial and the isolated. Only in cases where the back port change wouldn't isolated and difficult we should hold back on java8 usage.

martijnvg commented on plugins/ingest/src/test/java/org/elasticsearch/ingest/processor/AbstractStringProcessorTestCase.java:
```diff
@@ -0,0 +1,82 @@
+        }
+    }
+
+    public void testLowercaseNullValue() throws IOException {
```
maybe omit lowercase from the method names here? (since these tests also run for uppercase and trim)

javanna replied:
yea leftover...

martijnvg commented on plugins/ingest/src/test/java/org/elasticsearch/ingest/processor/lowercase/LowercaseProcessorFactoryTests.java:
```diff
@@ -0,0 +1,42 @@
+
+public class LowercaseProcessorFactoryTests extends ESTestCase {
+
+    public void testCreate() throws IOException {
```
maybe also add a test, where we expect failure if `fields` isn't specified?

javanna replied:
yes I am adding this test to every factory test.

martijnvg commented:
This is great! Left some minor comments. I do wonder if we group some processors in a single package? Right now there are a lot of packages with a single class, which kind of defeats the purpose of packages imo. Regardless LGTM.

javanna commented:
I rebased and addressed comments, also added missing javadocs to each of the new processor. I agree regarding packages, I followed the one processor package that we seem to have adopted, but we do end up with lots of packages, I guess we can discuss this as a followup.

martijnvg commented:
LGTM
